File: Swift-C++-Testsuite/test/typecasting/Swift/TypeInferences.swift
@main
struct MyMain {

    static func main() {
        let library = [
            Movie(name: "Casablanca", director: "Michael Curtiz"),
            Song(name: "Blue Suede Shoes", artist: "Elvis Presley"),
            Movie(name: "Citizen Kane", director: "Orson Welles"),
            Song(name: "The One And Only", artist: "Chesney Hawkes"),
            Song(name: "Never Gonna Give You Up", artist: "Rick Astley"),
        ]
        // the type of "library" is inferred to be [MediaItem]
        let movie = library[0]
        let song = library[1]
        let movieName = movie.name
        let songName = song.name
    }

}

class MediaItem {
    var name: String
    init(name: String) {
        self.name = name
    }
}

class Movie: MediaItem {
    var director: String
    init(name: String, director: String) {
        self.director = director
        super.init(name: name)
    }
}

class Song: MediaItem {
    var artist: String
    init(name: String, artist: String) {
        self.artist = artist
        super.init(name: name)
    }
}

File: Swift-C++-Testsuite/test/typecasting/C++/TypeInferences.cpp

#include <iostream> 
#include <string>
#include <vector>

using namespace std;

class MediaItem {
    public:
        string name; 
        MediaItem(string n) {
            name = n;
        }
};

class Movie : public MediaItem {
    public: 
        string director;
        Movie(string n, string d) : MediaItem(n) {
            director = d; 
        }
};

class Song : public MediaItem {
    public: 
        string artist;
        Song(string n, string a) : MediaItem(n) {
            artist = a; 
        }
};

int main() { 
    vector<MediaItem*> library;  
    library.push_back(new Movie("Casablanca", "Michael Curtiz")); 
    library.push_back(new Song("Blue Suede Shoes", "Elvis Presley")); 
    library.push_back(new Movie("Citizen Kane", "Orson Welles")); 
    library.push_back(new Song("The One And Only", "Chesney Hawkes"));
    library.push_back(new Song("Never Gonna Give You Up", "Rick Astley"));
     
    auto movie = library[0]; 
    auto song = library[1];
    auto movieName = movie->name; 
    auto songName  = song->name; 
    
    return 0; 
}